README.md
# xorg
xorg package


debian/changelog
xorg (1:7.7+19ubuntu7-galliumos0) bismuth; urgency=medium

  * Prepare for GalliumOS 3.0
    - patches from 1:7.7+13ubuntu3-galliumos1

 -- reynhout <reynhout@galliumos.org>  Mon, 07 May 2018 04:59:23 +0000

xorg (1:7.7+19ubuntu7) bionic; urgency=medium

  [ Simon Quigley ]
  * 60x11-common_xdg_path: If $XDG_CONFIG_DIRS is nonempty, don't discard the
    contents.

  [ Didier Roche ]
  * 60x11-common_xdg_path:
    - protect against multiple additions of the same paths to
      XDG_CONFIG_DIRS and XDG_DATA_DIRS, resulting to duplication on
      consecutive logout/logins due to older session hanging in and env
      being reused.
      Some consequences is that electron application crashes (LP: #1764355)
    - remove readdition of defaults if XDG_CONFIG_DIRS isn't empty to respect
      user's selection (same logic than wayland session).

 -- Timo Aaltonen <tjaalton@debian.org>  Thu, 19 Apr 2018 17:47:42 +0300

xorg (1:7.7+19ubuntu6) bionic; urgency=medium

  * control: Add python3-apport to xserver-xorg depends due to the apport hook.
    (LP: #1730035)
  * control: Depend on x11proto-dev instead of the old packages.

 -- Timo Aaltonen <tjaalton@debian.org>  Thu, 05 Apr 2018 15:47:54 +0300

xorg (1:7.7+19ubuntu5) bionic; urgency=medium

  * Cleanup upstart Xsession.d files, when upstart is removed, but not
    purged, and thus breaking graphical login after upgrades. LP: #1749199

 -- Dimitri John Ledkov <xnox@ubuntu.com>  Thu, 22 Feb 2018 16:43:18 +0000

xorg (1:7.7+19ubuntu4) bionic; urgency=medium

  * Recommend gnome-terminal before xterm to try to help germinate
    pick the correct alternate dependency (LP: #1720482)

 -- Jeremy Bicha <jbicha@ubuntu.com>  Tue, 30 Jan 2018 20:43:33 -0500

xorg (1:7.7+19ubuntu3) artful; urgency=medium

  * Add alternate dependency on gnome-terminal (LP: #1720482)

 -- Jeremy Bicha <jbicha@ubuntu.com>  Thu, 12 Oct 2017 10:25:30 -0400

xorg (1:7.7+19ubuntu2) artful; urgency=medium

  * control, source_xorg.py, xserver-xorg.install, rules: Import apport package
    hook and links from xdiagnose 3.8.7.
  * source_xorg.py: Drop obsolete stuff from the hook, attach Xorg.0.log
    from home directory.

 -- Timo Aaltonen <tjaalton@debian.org>  Thu, 31 Aug 2017 13:43:41 +0300

xorg (1:7.7+19ubuntu1) artful; urgency=medium

  * Merge from Debian.

 -- Timo Aaltonen <tjaalton@debian.org>  Wed, 03 May 2017 19:06:02 +0300

xorg (1:7.7+19) unstable; urgency=high

  * Team upload.
  * xserver-xorg now Recommends xserver-xorg-legacy, we'd have liked to avoid
    installing the setuid Xorg wrapper by default in stretch, but this seems
    to break too much still (closes: #861683).

 -- Julien Cristau <jcristau@debian.org>  Wed, 03 May 2017 11:38:16 +0200

xorg (1:7.7+18ubuntu1) artful; urgency=medium

  * Merge from Debian.
    - drop -synaptics from -input-all Recommends (LP: #1685542)

 -- Timo Aaltonen <tjaalton@debian.org>  Mon, 24 Apr 2017 16:38:55 +0300

xorg (1:7.7+18) unstable; urgency=medium

  * Call db_purge without arguments in postinst to fix error uncovered by
    adding "set -e" in 1:7.7+17 (closes: #845398).  Thanks, Samuel Thibault
    and Tianon Gravi!

 -- Julien Cristau <jcristau@debian.org>  Wed, 23 Nov 2016 08:50:50 +0100

xorg (1:7.7+17) unstable; urgency=medium

  [ Andreas Boll ]
  * control: Drop -freedreno from -video-all, the default 2D-driver will
    be -modesetting.

  [ Emilio Pozuelo Monfort ]
  * debian/x11-common.postinst: call `set -e'.
  * Bump Standards-Version to 3.9.8. No changes.
  * Bump debhelper compat to 10. Drop --parallel flag, enabled by default.

 -- Emilio Pozuelo Monfort <pochu@debian.org>  Wed, 23 Nov 2016 00:02:26 +0100

xorg (1:7.7+16ubuntu3) zesty; urgency=medium

  * control: Drop alt recommends on xorg-driver-video, because it makes
    no sense for Recommends.
  * control: Drop xserver-xorg-video-freedreno from video-all (arm).
    (LP: #1679323)

 -- Timo Aaltonen <tjaalton@debian.org>  Tue, 04 Apr 2017 00:47:23 +0300

xorg (1:7.7+16ubuntu2) zesty; urgency=medium

  * control: Add -synaptics to input-all Recommends for linux, and move
    wacom there.

 -- Timo Aaltonen <tjaalton@debian.org>  Thu, 20 Oct 2016 20:28:30 +0300

xorg (1:7.7+16ubuntu1) zesty; urgency=low

  * Merge from Debian unstable.  Remaining changes:
    - control: add xserver-xorg-video-geode for i386 Suggests
    - control: demote 'xserver-xorg-video-all | xorg-driver-video' to
      Recommends.
    - control: promote xserver-xorg-input-wacom to Depends
    - control: drop depend on libgl*-mesa* on armhf
    - control: add xinput to xorg Depends
    - control: demote xfonts-{100dpi,75dpi,scalable} * to xorg
      Recommends/Suggests
    - local/Xsession: truncate ERRFILE if it's too big
    - local/Xsession.d/20x11-common_process-args: Add support for
      session command containing args (LP: 512235)
    - local/Xsession.d/60x11-common_xdg_path: added to set
      xdg path depending on the selected session (compatible with
      lightdm, gdm and kdm now) (LP: 804734)
  * Dropped changes:
    - control: Drop freedreno and void from video/input-all.
  * Dropped changes, included in Debian:
    - control: Drop vmmouse from -input-all depends, our kernel has had
      the virtual driver for some time.
    - control: Demote -intel to -video-all Recommends, allowing uninstall
      without removing the metapackage.

 -- Steve Langasek <steve.langasek@ubuntu.com>  Tue, 18 Oct 2016 22:04:21 -0700

xorg (1:7.7+16) unstable; urgency=medium

  [ Timo Aaltonen ]
  * control: Drop -evdev and -synaptics from x-x-input-all Depends for
    linux.

  [ Julien Cristau ]
  * Drop xserver-xorg-input-void from s390x.  The server can come up with no
    input devices nowadays, -input-void isn't necessary.

  [ Cyril Brulebois ]
  * Drop xserver-xorg-input-vmmouse from xserver-xorg-input-all Depends,
    since the kernel driver for the VMware mouse device is incompatible
    with Xorg's vmmouse driver but works with the generic evdev and

debian/local/xserver-wrapper.c
/* xserver-wrapper.c - a simple wrapper for X servers that decides whether to
 * let them be run.
 *
 * By Stephen Early
 *
 * Stephen Early: modified to use /etc/X11/X symlink with security level of
 *                'Console' if /etc/X11/Xserver does not exist
 * Mark W. Eichin: permit non-privileged -showconfig (6 May 1997)
 * Mark W. Eichin: fix sense of error check for -showconfig (11 May 1997)
 * Mark W. Eichin: drop privileges on alternate -config, even if we do pass the
 *                 security check, to prevent using the error handling to read
 *                 the first line of any protected file (19 Sep 1997)
 * Erik Troan: prevent buffer overruns (25 Mar 1998)
 * Topi Miettinen: plug file descriptor leak (26 Apr 1998)
 * Branden Robinson: only fclose() if file was opened (3 May 1999)
 * Colin Phipps: minor device number check should be < 64, not < 128, or we
 *               catch serial terminals (26 Feb 2000)
 * Branden Robinson: ensure sanity of X server socket directory (13 Jun 2000)
 * Branden Robinson: make all paths #defines
 *                   more helpful socket dir error messages (29 Jun 2000)
 * Branden Robinson: bail out if the config file contains only the silly
 *                   default X server name (XF86_NONE) (30 Jul 2000)
 * Branden Robinson: increase verbosity when wrapper config not found
 *                   (2 Oct 2000)
 * Branden Robinson:
 *   - new configuration file, Xwrapper.config, with different format
 *     (name=value)
 *   - now just exec's /etc/X11/X; whatever this symlink points
 *     to will be used as the X server
 *   - config file specifies allowed user types as before (root only,
 *     console users, anyone)
 *   - config file specifies nice value to use for server
 *   (17 Nov 2000)
 * Branden Robinson: now accepts hyphens in variable contents (24 Nov 2000)
 * Branden Robinson: fix dumb errors left over from debugging (3 Dec 2000)
 * Branden Robinson: let root start the server even if he isn't on a
 *                    console, and the security level is console (11 Dec 2000)
 * Branden Robinson: check out the X server symlink with readlink; abort if
 *                   it's not a symlink, or if it points back to this wrapper
 *                   (24 Feb 2001)
 * Branden Robinson: whoops; readlink() doesn't null-terminate the target
 *                   string (27 Feb 20001)
 * Branden Robinson: add more info to "suspicious" error messages (16 Mar 2001)
 * Branden Robinson: also allow unprivileged use of "-version" option
 *                   (13 Jul 2001)
 * Branden Robinson: check mode of DRI device directory, if it exists, and warn
 *                   if it is weird (28 Aug 2001)
 * Branden Robinson: skip lines in Xwrapper.config that don't match expected
 *                   format (9 Dec 2001)
 * Branden Robinson: fix logic that was supposed to also allow unprivileged use
 *                   of "-version" option but which actually forbade both
 *                   "-showconfig" and "-verbose"; also let unprivileged users
 *                   specify "-help" option to get a usage message (26 Dec 2001)
 * Branden Robinson: change nice() usage to fit SuSv2 semantics; see Debian Bug
 *                   #140012 (2 Apr 2002)
 * Branden Robinson: *sigh* Ben Collins changed our FROZEN C library back to
 *                   pre-SuSv2 nice() semantics, so rewrote the nice() error
 *                   handling; also correct limits on legal nice values from
 *                   -20 <= x <= 20 to -20 <= x <= 19 (29 Apr 2002)
 * Branden Robinson: make the nice() error handling switchable with a #define
 *                   between SuSv2 semantics and old-style semantics
 *                   (16 Oct 2002)
 * Branden Robinson: stop using the GNU extension strnlen() to appease the
 *                   Debian GNU/NetBSD geeks, who are using BSD's C library
 *                   (16 Oct 2002)
 * Branden Robinson: chdir() to the directory where the X server symlink is kept
 *                   before executing its target, so that relative symlinks work
 *                   (1 Aug 2003)
 * Guillem Jover: add console detection support for GNU/kFreeBSD, and some
 *                messages at build and run time to allow the user to know
 *                what failed on unsupported systems
 *                (30 Mar 2007)
 * Brice Goglin: drop privileges on alternate config file given with
 *               -xf86config (14 Jun 2007)
 * Tollef Fog Heen: stop handling -config specifically, since newer
 *                  Xorg does that check automatically and only allows
 *                  non-root users to specify configuration files
 *                  relative to /etc/X11 (10 Aug 2007)
 * Loïc Minier: on Linux, also consider alternate tty devices (major 5 and
 *              minor < 64) as consoles (24 Sep 2008)
 * Julien Cristau: remove the nice_value option
 * Julien Cristau: recognize /usr/bin/X as a path to this wrapper (6 Jun 2009)
 * Julien Cristau: don't print an error message if Xwrapper.config doesn't exist
 *                 (11 Aug 2009)
 * Julien Cristau: allow unprivileged -showDefaultModulePath and
 *                 -showDefaultLibPath options (11 Aug 2009)
 * Julien Cristau: don't check the mode of the DRI device directory
 *                 (11 Aug 2009)
 * Julien Cristau: also drop group privileges (1 Nov 2011)
 * Julien Cristau: disallow major 5 again for consoles (15 Dec 2011)
 *
 * This is free software; you may redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2,
 * or (at your option) any later version.
 *
 * This is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License with
 * the Debian operating system, in /usr/share/common-licenses/GPL;  if
 * not, write to the Free Software Foundation, Inc., 59 Temple Place,
 * Suite 330, Boston, MA 02111-1307 USA
 *
 */

#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/types.h>

#if defined(__linux__)
#define TTY_MAJOR_DEV 4
#elif defined(__FreeBSD__) || defined(__FreeBSD_kernel__)
#include <sys/consio.h>
#endif

#define X_WRAPPER_CONFIG_FILE "/etc/X11/Xwrapper.config"
#define X_SERVER_SYMLINK_DIR "/etc/X11"
#define X_SERVER_SYMLINK "/etc/X11/X"
#define X_SOCKET_DIR "/tmp/.X11-unix"
#define X_SOCKET_DIR_MODE (S_ISVTX | S_IRWXU | S_IRWXG | S_IRWXO)

#ifndef FALSE
#define FALSE 0
#endif
#ifndef TRUE
#define TRUE 1
#endif

typedef enum {
  RootOnly,
  Console,
  Anybody
} SecurityLevel;

static SecurityLevel
getSecLevel(char *security)
{
  char *c;

  for (c = security; *c; c++) *c = toupper(*c);

  if (strncmp(security,"ROOTONLY",8) == 0) return RootOnly;
  if (strncmp(security,"CONSOLE",7) == 0) return Console;
  if (strncmp(security,"ANYBODY",7) == 0) return Anybody;
  return RootOnly;
}

static int
onConsole()
{
#if defined(__linux__)
  struct stat s;

  /* see if stdin is a virtual console device */
  if (fstat(0, &s) != 0) {
    (void) fprintf(stderr, "X: cannot stat stdin\n");
    return FALSE;
  }
  if (S_ISCHR(s.st_mode) &&
        (major(s.st_rdev) == TTY_MAJOR_DEV &&
         minor(s.st_rdev) < 64)) {
    return TRUE;
  }
#elif defined(__FreeBSD__) || defined(__FreeBSD_kernel__)
  int idx;

  if (ioctl(0, VT_GETINDEX, &idx) != -1)
    return TRUE;
#else
#warning This program needs porting to your kernel.
  (void) fprintf(stderr, "X: unable to determine if running on a console\n");
#endif

  return FALSE;
}

static int
checkSecLevel(SecurityLevel level)
{
  switch (level) {
  case RootOnly:
    if (getuid() == 0) { /* real uid is root */
      return TRUE;
    } else {
      return FALSE;
    }
    break;
  case Console:
    if (getuid() == 0) return TRUE; /* root */
    return onConsole();
    break;
  case Anybody:
    return TRUE;
  }
  return FALSE;
}

int
main(int argc, char **argv)
{
  FILE *cf;
  struct stat statbuf;
  char xserver[1025];
  char line[1024];
  char var[65];
  char value[257];
  int length;
  int i;
  char *val;
  mode_t mask;
  SecurityLevel level = RootOnly;

  /* attempt to use our config file */
  cf = fopen(X_WRAPPER_CONFIG_FILE, "r");

  if (cf) {
    /* parse it */

    val = fgets(line, 1024, cf);

    while (val != NULL) {
      var[0] = '\0';
      value[0] = '\0';
      if (sscanf(line, " %64[A-Za-z0-9_] = %256[A-Za-z0-9_ -] ",
                 var, value) > 0) {
        /* truncate extra spaces at end of value */
        length = strlen(value);
        if (length > 256) {
          length = 256;
        }
        for (i = (length - 1); (value[i] == ' '); i--) {
          value[i] = '\0';
        }
        /* DEBUG (void) fprintf(stderr, "var: %s, value: %s.\n", var, value); */
        if (strncasecmp(var, "allowed_users", 64) == 0) {
          level = getSecLevel(value);
          /* DEBUG (void) fprintf(stderr, "security level set to %d\n", level); */
        }
      }
      val = fgets(line, 1024, cf);
    }

    (void) fclose(cf);
  } else {
    /* DEBUG (void) fprintf(stderr, "X: unable to open wrapper config file %s\n",
                   X_WRAPPER_CONFIG_FILE); */
  }

  i = lstat(X_SERVER_SYMLINK, &statbuf);
  if (i < 0 && errno != ENOENT) {
    (void) fprintf(stderr, "X: cannot stat %s (%s), aborting.\n",
                   X_SERVER_SYMLINK, strerror(errno));
    exit(1);
  }

  if (i < 0) {
    strcpy(xserver, "/usr/bin/Xorg");
  } else {
    i = readlink(X_SERVER_SYMLINK, xserver, 1024);

    if (i < 0) {
      (void) fprintf(stderr, "X: cannot read %s symbolic link (%s), aborting.\n",
                     X_SERVER_SYMLINK, strerror(errno));
      exit(1);
    }
    xserver[i] = '\0'; /* readlink() does not null-terminate the string */

    if ((strcmp(xserver, "/usr/bin/X11/X") == 0) ||
        (strcmp(xserver, "/usr/X11R6/bin/X") == 0) ||
        (strcmp(xserver, "/usr/bin/X") == 0)) {
      (void) fprintf(stderr, "X: %s points back to X wrapper executable, "
                     "aborting.\n", X_SERVER_SYMLINK);
      exit(1);
    }
  }

  if (access(xserver, X_OK)) { /* access() uses real uid */
    (void) fprintf(stderr, "%s is not executable\n", val);
    exit(1);
  }

  /* do we have permission to run the X server? */
  if (checkSecLevel(level)) {
    /* check for a sane server socket dir */
    mask = umask(0);
    /* some stupid kernels can't set the sticky bit during a mkdir() */
    if (!(mkdir(X_SOCKET_DIR, X_SOCKET_DIR_MODE))) {
      (void) chmod(X_SOCKET_DIR, X_SOCKET_DIR_MODE);
    }
    (void) umask(mask);

    /* do paranoid checks on the directory where the X server creates its socket */
    if (lstat(X_SOCKET_DIR, &statbuf)) {
      (void) fprintf(stderr, "X: cannot stat %s (%s), aborting.\n",
                     X_SOCKET_DIR, strerror(errno));
      exit(1);
    }

    if ((statbuf.st_uid != 0) || (statbuf.st_gid != 0)) {
      (void) fprintf(stderr, "X: %s has suspicious ownership (not root:root), "
                     "aborting.\n", X_SOCKET_DIR);
      exit(1);
    }

    if (statbuf.st_mode != (S_IFDIR | X_SOCKET_DIR_MODE)) {
      (void) fprintf(stderr, "X: %s has suspicious mode (not %o) or is not a "
                     "directory, aborting.\n", X_SOCKET_DIR, X_SOCKET_DIR_MODE);
      exit(1);
    }

    for (i = 1; i < argc; i++) {
      if (strlen(argv[i]) > 256) {
        if (setgid(getgid()) || setuid(getuid())) {
          perror("X unable to drop setuid privileges for suspiciously long "
                 "argument");
          exit(1);
        }
      }
    }

    /* run the X server */
    seteuid(0);

    /* DEBUG exit(0); */

    /*
     * change to the directory where the X server symlink is so that a relative
     * symlink will work and execute the X server
     */
    if (xserver[0] != '/' && chdir(X_SERVER_SYMLINK_DIR)) {
      (void) fprintf(stderr, "X: cannot chdir() to %s (%s), aborting.\n",
                     X_SERVER_SYMLINK_DIR, strerror(errno));
      exit(1);
    }
    (void) execv(xserver, argv);
    (void) fprintf(stderr, "X: exec of %s failed\n", xserver);
    exit(1);

  } else {
      /* DEBUG fprintf(stderr, "argc = %d, argv[1] = \"%s\"\n", argc, argv[1]); */
      /* DEBUG fprintf(stderr, "strcmp(argv[1], \"-showconfig\") = %d, strcmp(argv[1],
        \"-version\" = %d\n", (strcmp(argv[1], "-showconfig")), (strcmp(argv[1],
        "-version"))); */
      if (argc == 2 && ( (strcmp(argv[1], "-help") == 0) ||
                         (strcmp(argv[1], "-showconfig") == 0) ||
                         (strcmp(argv[1], "-version") == 0) ||
                         (strcmp(argv[1], "-showDefaultModulePath") == 0) ||
                         (strcmp(argv[1], "-showDefaultLibPath") == 0) ) ) {
          if (setgid(getgid()) || setuid(getuid())) {
              perror("X unable to drop setuid privileges");
              exit(1);
          }
          execv(xserver,argv);
          (void) fprintf(stderr, "X: unprivileged exec of %s failed, "
                         "aborting.\n", xserver);
          exit(1);
      } else {
          (void) fprintf(stderr, "X: user not authorized to run the X "
                         "server, aborting.\n");
          exit(1);
      }
  }

  (void) fprintf(stderr, "X: Impossible!  Unreachable statement reached!\n");
  exit(1);
}

/*
 * vim:set cindent et fo=tcroq sts=2 sw=2 tw=80:
 */
